### ISSUE TITLE
output include file content inside an included template

### ISSUE THREAD

**Issue by mindon** [caddyserver/caddy#4721]
here's the **partials.html** template

```
{{$stats := (include "/abc.json" | splitFrontMatter)}}
{{$keys := split "," "a,b"}}
{{$x := $stats.Meta}}
{{ range $keys }}
  {{.}} = {{ index $x . }}
{{ end }}

```

content of **abc.json** is
```
{
  "a": 1,
  "b": 2,
  "c": 3
}

```

if visiting **partials.html** directly, everything is ok

but if there's a **hell.html** , something wired shows up

`{{ include "/partials.html" }}`

sometimes it prints out `{ "a": 1, "b": 2, "c": 3 } a = 1 b = 2`  instead of `a = 1 b = 2`

if `(include "/abc.json" | splitFrontMatter)` is replaced with `fromJson "{ \"a\": 1, \"b\": 2, \"c\": 3 }"`, everything is ok

**Comment by mholt** [comment]
Thanks for opening an issue! We'll look into this.

It's not immediately clear to me what is going on, so I'll need your help to understand it better.

Ideally, we need to be able to reproduce the bug _in the most minimal way possible_. This allows us to write regression tests to verify the fix is working. If we can't reproduce it, then you'll have to test our changes for us until it's fixed -- and then we can't add test cases, either.

I've attached a template below that will help make this easier and faster! **This will require some effort on your part -- please understand that we will be dedicating time to fix the bug you are reporting if you can just help us understand it and reproduce it easily.**

This template will ask for some information you've already provided; that's OK, just fill it out the best you can. :+1: I've also included some helpful tips below the template. Feel free to let me know if you have any questions!

Thank you again for your report, we look forward to resolving it!

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
paste here
```


### 1c. Go version (if building Caddy from source; run `go version`)

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Instructions -- please heed otherwise we cannot help you (help us help you!)

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by mindon** [comment]
the origin description is misleading to (include | splitFrontMatter), but the problem is raised up with a include inside a included template, a very simple case is like following:

a.html  `{{ include "/b.html" }}`

b.html `{{ $c := (include "/anything") }}`

if visiting a.html, it will output the content of **/anything**.  

have no idea if it is a wrong use case with **include** or a bug...

**Comment by mindon** [comment]
it's surely a include bug inside a include template

the simplest case i mentioned above, everything output in **b.html** before the **include** will lost due to the code  **buf.Reset()** around line 214 inside **executeTemplateInBuffer** in file 
 https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/templates/tplcontext.go

**Comment by francislavoie** [comment]
This is likely related to changes in https://github.com/caddyserver/caddy/pull/4321, if @rockorager could take a look at this issue?

**Comment by rockorager** [comment]
Yep, I’ll look into it!

**Comment by rockorager** [comment]
I'm not able to reproduce. Can you list the exact steps you are doing to create the issue and the version/build you are running?

For example, I am doing this:
index.html
```
A
{{ include "b.html" }}
```
b.html
```
pre-include B
{{ include "c.html"  }}
{{ $d := (include "d.html") }}
post-include B
```
c.html
```
C
```
d.html
```
D
```

Visit index.html:

Output:
```
A pre-include B C post-include B
```

**Comment by mindon** [comment]
@rockorager 
**MacBook Pro M1 (arm64)**

**caddy version**
v2.4.6 h1:HGkGICFGvyrodcqOOclHKfvJC0qTU7vny/7FhYp9hNw=

with your code, got following 3 output cases when keep refreshing in Chrome
case 1 `A pre-include B C post-include B`

case 2 `A D post-include B`

case 3 `A CC post-include B`

then i try with following testing code in devtools console,

```javascript
const s={};
for(let i=0; i<1000; i++) {
    const r = await fetch('http://localhost/index.html')
      .then(resp => resp.text())
    s[r]=(s[r]||0)+1;
}
console.log(JSON.stringify(s, '', '\t'));

```
the report would be something like this

```json
{
	"A\nD\npost-include B": 680,
	"A\npre-include B\nC\n\npost-include B": 291,
	"A\nCC\n\npost-include B": 29
}

```

and if i try following

```javascript
const s={};
for(let i=0; i<1000; i++) {
    fetch('http://localhost/index.html')
      .then(resp => resp.text())
      .then(r=>s[r]=(s[r]||0)+1);
}

```

when all request done,  run 
```javascript
console.log(JSON.stringify(s, '', '\t'))

```
here's the result

```json
{
	"A\nD\npost-include B": 645,
	"A\npre-include B\nC\n\npost-include B": 117,
	"A\nC\npost-include B": 8,
	"A\npre-include B\nC\npost-include B\n\npost-include B": 3,
	"A\nCC\n\npost-include B": 86,
	"A\nC": 7,
	"A\nD": 11,
	"A\nD\npost-include B\npost-include BD\npost-include B\npost-include B\n\npost-include B": 1,
	"A\n": 12,
	"A\nA\n\npost-include B": 5,
	"A\n\npost-include BD\npost-include B": 5,
	"A\npre-include B\n": 8,
	"A\npre-include B\nC\n\npost-include Bpre-include B\nC\n\npost-include B\n\npost-include B": 2,
	"A\nDD\n\npost-include B": 3,
	"A\nDCCC\n\npost-include B\n\npost-include B": 1,
	"A\npre-include B\n\npost-include B": 5,
	"A\nDC\n\npost-include B": 5,
	"A\nD\npost-include BC\n": 2,
	"A\npre-include B\n{{ include \"c.html\"  }}\n{{ $d := (include \"d.html\") }}\npost-include B\npost-include B": 10,
	"A\nD\npost-include B\npost-include B": 10,
	"A\npre-include B\npre-include B\n\n\npost-include B": 1,
	"A\nA\nD\npost-include B": 1,
	"A\nD\npost-include BD\npost-include B": 1,
	"A\nA\nA\n\n\npost-include B": 1,
	"A\nA\n\n\npost-include Bpre-include B\n{{ include \"c.html\"  }}\n{{ $d := (include \"d.html\") }}\npost-include B\npost-include B": 1,
	"A\npre-include B\nD\n\npost-include B": 1,
	"pre-include B\n": 1,
	"A\nC\n\npost-include B": 1,
	"A\nD\npost-include B\npost-include B\npost-include B": 1,
	"A\nA\nD\npost-include B\npost-include B\npost-include B": 1,
	"": 1,
	"A\nA\n\nD\npost-include B": 2,
	"A\n\npost-include B": 5,
	"A\npre-include B\nCC\n\n\npost-include B": 2,
	"A\npre-include B\nCC\n\npost-include B\n\npost-include B": 2,
	"A\nD\npost-include Bpre-include B\nC\n\npost-include B": 1,
	"A\nA\npre-include B\n\npost-include BA\npre-include B\n\npost-include B\n\npost-include B": 1,
	"A\npre-include B\n\npost-include BA\npre-include B\n\npost-include B\n\npost-include BD\npost-include BC\n": 1,
	"A\nCC\n\npre-include B\n{{ include \"c.html\"  }}\n{{ $d := (include \"d.html\") }}\npost-include B\npost-include B": 1,
	"A\npre-include B\n\npost-include BC\n\npost-include B": 1,
	"A\npre-include B\nC\nC\n\npost-include B\npost-include B": 2,
	"A\npre-include B\nC\nC\n\npost-include B": 2,
	"A\nA\n{{include \"b.html\"}}C\n\npost-include B": 1,
	"A\nD\npost-include BC\n\npost-include B": 2,
	"A\nDpre-include B\nC\n\npost-include B": 1,
	"\npost-include BA\nD\npost-include B": 2,
	"A\nD\npost-include BD\npost-include B\n\npost-include B\npost-include B": 1,
	"A\nD\npost-include BD\npost-include B\n\npost-include B": 3,
	"pre-include B\nC\nD\npost-include B": 1,
	"A\nA\n": 1,
	"A\nC\nA\n{{include \"b.html\"}}\npost-include B": 1,
	"C\nA\nD\npost-include B\npost-include B": 1,
	"A\nA\n{{include \"b.html\"}}\npost-include B": 1,
	"A\nDpre-include B\nC\n\npost-include B\n\npost-include B": 1,
	"A\npre-include B\nA\n\n\npost-include B": 1,
	"A\npre-include B\npre-include B\nC\n\npost-include B\n\npost-include B": 1,
	"A\nA\nA\n\npost-include B": 1,
	"A\n\npost-include Bpre-include B\nC\n\npost-include Bpre-include B\nC\n\npost-include B\n\npost-include BC\n": 1,
	"A\nA\n\npost-include Bpre-include B\nC\n\npost-include B": 1,
	"A\n\n\npost-include B": 1,
	"A\n\n\npost-include B\npost-include B": 1
}

```

**Comment by rockorager** [comment]
@mindon thanks for the details.

This should already be fixed in the upcoming 2.5.0 RC. A similar bug (#4423) was reported for 2.4.6 as well and fixed [here](#4452).

Are you able to reproduce on 2.5.0?

**Comment by mindon** [comment]
@rockorager  thank you very much, it works great with caddy v2.5.0-rc.1 👍
